core/liquidity/v2/sla.go

```diff
@@ -0,0 +1,229 @@
+	if e.auctionState.InAuction() {
+		minPriceFactor := num.Min(e.orderBook.GetLastTradedPrice(), e.orderBook.GetIndicativePrice()).ToDecimal()
+		maxPriceFactor := num.Max(e.orderBook.GetLastTradedPrice(), e.orderBook.GetIndicativePrice()).ToDecimal()
+
+		// (1.0-market.liquidity.priceRange) x min(last trade price, indicative uncrossing price)
+		minPrice = one.Sub(e.priceRange).Mul(minPriceFactor)
+		// (1.0+market.liquidity.priceRange) x max(last trade price, indicative uncrossing price)
+		maxPrice = one.Add(e.priceRange).Mul(maxPriceFactor)
+	} else {
+		mid, err := e.getMidPrice()
+		// if there is no mid price then LP is not meeting their committed volume of notional.
+		if err != nil || mid.IsZero() {
+			return false
+		}
+
+		midD := mid.ToDecimal()
+		// (1.0 - market.liquidity.priceRange) x mid
+		minPrice = one.Sub(e.priceRange).Mul(midD)
+		// (1.0 + market.liquidity.priceRange) x mid
+		maxPrice = one.Add(e.priceRange).Mul(midD)
+	}
+
+	notionalVolume := num.DecimalZero()
+	orders := e.getAllActiveOrders(party)
+
+	for _, o := range orders {
+		price := o.Price.ToDecimal()
+		// this order is in range and does contribute to the volume on notional
+		if price.GreaterThanOrEqual(minPrice) && price.LessThanOrEqual(maxPrice) {
+			notionalVolume = notionalVolume.Add(price)
+		}
+	}
+
+	requiredLiquidity := e.stakeToCcyVolume.Mul(lp.CommitmentAmount.ToDecimal())
+	return notionalVolume.GreaterThanOrEqual(requiredLiquidity)
+}
+
+func (e *Engine) calculateCurrentFeePenalty(timeBookFraction num.Decimal) num.Decimal {
+	one := num.DecimalOne()
+
+	// p = (1-[timeBookFraction-commitmentMinTimeFraction/1-commitmentMinTimeFraction]) * slaCompetitionFactor
+	return one.Sub(
+		timeBookFraction.Sub(e.commitmentMinTimeFraction).Div(one.Sub(e.commitmentMinTimeFraction)),
+	).Mul(e.slaCompetitionFactor)
+}
+
+func (e *Engine) calculateBondPenalty(timeBookFraction num.Decimal) num.Decimal {
+	// min(nonPerformanceBondPenaltyMax, nonPerformanceBondPenaltySlope * (1-timeBookFraction/commitmentMinTimeFraction))
+	min := num.MinD(
+		e.nonPerformanceBondPenaltyMax,
+		e.nonPerformanceBondPenaltySlope.Mul(num.DecimalOne().Sub(timeBookFraction.Div(e.commitmentMinTimeFraction))),
+	)
+
+	// max(0, min)
+	return num.MaxD(num.DecimalZero(), min)
+}
+
+func (e *Engine) calculateHysteresisFeePenalty(currentPenalty num.Decimal, previousPenalties []num.Decimal) num.Decimal {
+	l := len(previousPenalties)
+	if l < 1 {
+		return currentPenalty
+	}
+
+	// Select window windowStart for hysteresis period
+	windowStart := l - int(e.performanceHysteresisEpochs)
+	if windowStart < 0 {
+		windowStart = 0
+	}
+
+	periodAveragePenalty := num.DecimalZero()
+	for _, p := range previousPenalties[windowStart:] {
+		periodAveragePenalty = periodAveragePenalty.Add(p)
+	}
+
+	performanceHysteresisEpochsD := num.NewDecimalFromFloat(float64(e.performanceHysteresisEpochs))
+
+	periodAveragePenalty = periodAveragePenalty.Div(performanceHysteresisEpochsD)
```

Comment:
Wondering if we should be dividing by `len(previousPenalties[windowStart:])` here so thats its the average of whatever history we have. For example if this is only the second epoch for an LP, they had a penatly of `1` in the previous epoch and `performanceHysteresisEpochs = 10` we'll get their historic average being `1/10`.

This could just be my intrepretation of the spec though.